### PR TITLE
[Experimental] Remove store barrier after PG init

### PR DESCRIPTION
**Pull request by kwen2501**
Store based barrier is not scalable.
Experimenting to see if removing it breaks any CI

### CODE REVIEW

**Review by rohan-varma — COMMENTED** [review]
Looks like CI is fine, but I think we may still need some sort of barrier? See https://github.com/pytorch/pytorch/pull/49419, https://github.com/pytorch/pytorch/issues/48110 for some prior context. If we don't have a barrier, it might be tough to figure out where is the actual synchronization point before training commences which is useful for debugging slow / stuck ranks?

**Comment by kwen2501** [comment]
Thanks @rohan-varma for the quick review :)
That's why this PR is experimental.
But now that all CI's are green, I plan to add an environment variable to activate this removal when the env is set.

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -957,16 +964,27 @@ def init_process_group(
...
-    if backend == Backend.MPI:
-        # MPI backend doesn't use store.
-        barrier()
+    if _barrier_after_init == 1:
+        # barrier at the end to ensure that once we return from this method, all
+        # process groups including global variables are updated correctly on all
+        # ranks.
+        # Update 04/2023: for large-scale runs, this barrier (esp. store-based
+        # barrier) may be costly and/or unscalable. Also, in a lot of cases,
+        # these barriers may be unnecessary, as proved by a green CI after
```
Comment (kumpera) [inline torch/distributed/distributed_c10d.py:973]: I don't think we can say just because CI is green that it's ok to remove the barrier.
You didn't even run the multi-gpu tests so the odds of running into the actual issues is slim to none.

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -468,6 +468,13 @@ def _get_pg_device(group: ProcessGroup):
     return torch.device("cpu")
 
 
+# Environment variable to control whether we do a barrier after process group
+# init. Default value is 1 for now to stay the same with previous behavior.
+# Users can change it to 0 if such behavior is undesired. We reserve the right
+# to change the default value to 0 if small rollout is successful.
+_barrier_after_init = int(os.getenv("TORCH_DIST_INIT_BARRIER", "1"))
```
Comment (kumpera) [inline torch/distributed/distributed_c10d.py:475]: Could you update the docs as well to surface this new knob?

**Review by kumpera — APPROVED** [review]
I guess it's worth trying removing the barrier and see what breaks.

File: torch/distributed/distributed_c10d.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/distributed_c10d.py:973]: Added the periodic-ci label.

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -3603,16 +3621,27 @@ def _new_group_with_tag(ranks=None, timeout=default_pg_timeout, backend=None, pg
...
+        # these barriers may be unnecessary, as proved by a green CI after
+        # removal. An environment variable `TORCH_DIST_INIT_BARRIER` has been
+        # added which, when set to 0, will disable these barriers.
+        if backend == Backend.MPI:
+            # MPI doesn't have store.
+            barrier()
+        else:
+            # Use store based barrier here since barrier() used a bunch of
+            # default devices and messes up NCCL internal state.
+            _store_based_barrier(global_rank, default_store, timeout)
```
Comment (H-Huang) [inline torch/distributed/distributed_c10d.py:3639]: I believe that this is the only place that `_store_based_barrier` is used and its not a public api, so if we are guarding it here with the `_barrier_after_init` flag, we also don't need to make any logic changes in the function right?

**Review by H-Huang — APPROVED** [review]
LGTM, we can try it out since default behavior remains the same, thanks for adding the flag

File: torch/distributed/distributed_c10d.py [hunk 4]
```diff
@@ -3603,16 +3621,27 @@ def _new_group_with_tag(ranks=None, timeout=default_pg_timeout, backend=None, pg
...
     else:
-        # Use store based barrier here since barrier() used a bunch of
-        # default devices and messes up NCCL internal state.
-        _store_based_barrier(global_rank, default_store, timeout)
+        logger.info(
+            "TORCH_DIST_INIT_BARRIER is set to 0, omitting the barrier after "
+            "ProcessGroup initialization."
+        )
```
Comment (fduwjj) [inline torch/distributed/distributed_c10d.py:3644]: Does this mean we call barrier twice because `_new_process_group_helper` is called in `init_process_group` ?

Comment (kwen2501) [reply torch/distributed/distributed_c10d.py:3644]: Yes. Super sad!

File: torch/distributed/distributed_c10d.py (hunk 3 above)
Comment (kwen2501) [reply torch/distributed/distributed_c10d.py:3639]: > we also don't need to make any logic changes in the function right?
If by "the function" you mean `_store_based_barrier`, then I believe so.

**Comment by pritamdamania87** [comment]
More context about this can also be found in https://github.com/pytorch/pytorch/pull/45181

**Comment by kwen2501** [comment]
Thanks for the context @pritamdamania87 !

You mentioned in #45181 that:
> As a result, there is a race that after initializing the process group on say rank 0, if we immediately
check the default process group on rank 1 (say via RPC), we might actually get
an error since rank 1 hasn't yet updated its _default_pg variable.

I wonder how prevalent is such case now. Even if it exists, I think this is an issue of RPC -- its programming model implies a need for synchronization between caller and callee. So this is not a problem of `init_process_group`, RPC may hit the same issue with other calls too. Hence I think the synchronization should be made on the RPC side, rather than from c10d (the store-based barrier is just doing a favor here.)

The reason for removing the store-based barrier is that it is unfortunately unscalable. Large jobs today fail because of it.

If the previous concern was mainly from RPC, I am actually more confident in making the removal default, because the described scenario does not seem like something c10d should support.

**Comment by pritamdamania87** [comment]
> If the previous concern was mainly from RPC, I am actually more confident in making the removal default, because the described scenario does not seem like something c10d should support.

@kwen2501  I don't think this is only for RPC but in general any other mechanism that could inspect the state of the process groups externally. Removing the barrier by default seems dangerous to me and could lead hard to debug issues further down in the program. In terms of a general API contract it seems sound to have a barrier which ensures all ranks have successfully process `init_process_group` before continuing. For example, if you remove the barrier some ranks could die in the middle of init_process_group and other ranks might continue their program. This could lead to weird hangs or other issues  further along in the program which are hard to debug.

Regarding scalability issues, could you elaborate a bit more about what is the exact issue? Typically `init_process_group` is called only once during the lifetime of the program and is typically fine to incur the overhead of a barrier. Is the use case  you are worried about where we have several short running programs that need to call `init_process_group` or does a long running program called `init_process_group` multiple times?

**Comment by kwen2501** [comment]
Thanks @pritamdamania87 . It is not about the overhead of the barrier. The TCPStore can just error out at that large scale.

**Comment by pritamdamania87** [comment]
> Thanks @pritamdamania87 . It is not about the overhead of the barrier. The TCPStore can just error out at that large scale.

@kwen2501 The TCPStore is anyways used during the initialization (ex: sharing ncclUniqueId across all ranks) so I'm not sure how just removing the store based barrier would help? The fix should really be making the TCPStore scalable.

**Comment by wdykas** [comment]
@pritamdamania87 Just to add another perspective in our case its not about erroring out but about start up time of large jobs.  The hope is that by removing the barrier large jobs will start up faster without having to hit a barrier.

**Comment by pritamdamania87** [comment]
@wdykas Thanks for sharing the perspective. Was wondering how long do your large jobs run and how much is the start time from the barrier? Typically large jobs run for a long time and some additional startup time shouldn't hurt much?

Despite that, I think having a flag to skip the store based barrier makes sense. But I'm more worried about removing the store based barrier completely or making the default to skip it.
